packages/contracts/src/core/permission/PermissionManager.sol

```diff
@@ -215,16 +220,96 @@ abstract contract PermissionManager is Initializable {
         bytes32 _permissionId,
         bytes memory _data
     ) public view virtual returns (bool) {
-        return
-            _isGranted(_where, _who, _permissionId, _data) || // check if `_who` has permission for `_permissionId` on `_where`
-            _isGranted(_where, ANY_ADDR, _permissionId, _data) || // check if anyone has permission for `_permissionId` on `_where`
-            _isGranted(ANY_ADDR, _who, _permissionId, _data); // check if `_who` has permission for `_permissionI` on any contract
+        // Specific caller and target
+        {
+            address specificPermissionOrCondition = permissionsHashed[
+                permissionHash({_where: _where, _who: _who, _permissionId: _permissionId})
+            ];
+
+            if (specificPermissionOrCondition == ALLOW_FLAG) return true;
+
+            if (specificPermissionOrCondition != UNSET_FLAG) {
+                return
+                    _checkCondition({
+                        _condition: specificPermissionOrCondition,
+                        _where: _where,
+                        _who: _who,
+                        _permissionId: _permissionId,
+                        _data: _data
+                    });
+            }
+        }
+
+        // Generic caller `_who: ANY_ADDR`  (only possible by granting with `grantWithCondition`)
+        {
+            address genericCallerCondition = permissionsHashed[
+                permissionHash({_where: _where, _who: ANY_ADDR, _permissionId: _permissionId})
+            ];
+
+            if (genericCallerCondition != UNSET_FLAG) {
+                return
+                    _checkCondition({
+                        _condition: genericCallerCondition,
+                        _where: _where,
+                        _who: _who,
+                        _permissionId: _permissionId,
+                        _data: _data
+                    });
+            }
+        }
```

Comment:
this part is duplicated. wouldn't it make sens to move to its own function where you only pass the 3 parameters?

Reply:
or move it into `_checkCondition` and rename the function to `_checkPermission` which includes the condition part

Reply:
I think the problem is that the function would need to return `bool` but in some cases, we don't want to return but move to the next code part.
Do I overlook something?